src/SSHDebugPS/RawShell.cs

```diff
@@ -0,0 +1,295 @@
+    {
+        event EventHandler<string> OutputReceived;
+        event EventHandler<int> Closed;
+        event EventHandler ErrorOccured;
+
+        void WriteCommandStart(string startCommand);
+        void Write(string text);
+        void WriteLine(string text);
+    }
+
+    internal class RawLocalShell : IRawShell
+    {
+        private System.Diagnostics.Process _localProcess;
+        private StreamWriter _stdoutWriter;
+
+        public RawLocalShell(string command, string arguments)
+        {
+            ProcessStartInfo processStartInfo = new ProcessStartInfo(command, arguments);
+            processStartInfo.RedirectStandardError = true;
+            processStartInfo.RedirectStandardInput = true;
+            processStartInfo.RedirectStandardOutput = true;
+            processStartInfo.UseShellExecute = false;
+            processStartInfo.CreateNoWindow = true;
+
+            _localProcess = System.Diagnostics.Process.Start(processStartInfo);
+            _stdoutWriter = new StreamWriter(_localProcess.StandardInput.BaseStream, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false), 4096, leaveOpen: true);
+            _localProcess.OutputDataReceived += OnProcessOutput;
+            _localProcess.ErrorDataReceived += OnErrorOutput;
+            _localProcess.Exited += OnProcessExited;
+
+            _localProcess.BeginOutputReadLine();
+            _localProcess.BeginErrorReadLine();
+            _localProcess.EnableRaisingEvents = true;
+        }
+
+        public event EventHandler<string> OutputReceived;
+        public event EventHandler<int> Closed;
+        public event EventHandler ErrorOccured;
+
+        public void Dispose()
+        {
+            if (_localProcess != null)
+            {
+                _localProcess.Close();
+            }
+
+            if (_stdoutWriter != null)
+            {
+                _stdoutWriter.Close();
+                _stdoutWriter = null;
+            }
+        }
+
+        public void Write(string text)
+        {
+            _stdoutWriter.Write(text);
+            _stdoutWriter.Flush();
+        }
+
+        public void WriteCommandStart(string startCommand)
+        {
+            WriteLine(startCommand);
+        }
+
+        public void WriteLine(string text)
+        {
+            _stdoutWriter.WriteLine(text);
+            _stdoutWriter.Flush();
+        }
+
+        private void OnErrorOutput(object sender, DataReceivedEventArgs e)
+        {
+            ErrorOccured?.Invoke(sender, e);
+        }
+
+        private void OnProcessOutput(object sender, DataReceivedEventArgs e)
+        {
+            OutputReceived?.Invoke(this, e?.Data + '\n');
+        }
+
+        private void OnProcessExited(object sender, EventArgs e)
+        {
+            Closed?.Invoke(this, _localProcess.ExitCode);
+        }
+    }
+
+    /// <summary>
+    /// Launches a raw command that doesn't have shell-like input/output
+    /// </summary>
+    internal class CommandShell : IRawShell
```

Comment:
>CommandShell [](start = 19, length = 12)

Is the difference between this and RawLocalShell that RawLocalShell only provides full lines as output, and this will also provide data that doesn't end with a newline?

Reply:
This one doesn't do readline and uses a readloop witha buffer instead. basically avoiding the readline implementation for the above.

---
In reply to: [254097210](https://github.com/Microsoft/MIEngine/pull/831#discussion_r254097210) [](ancestors = 254097210)

Reply:
I need a better name for these two classes but i don't know what to call it. The other option is to allow a boolean and implement both in the same class?

---
In reply to: [256226539](https://github.com/Microsoft/MIEngine/pull/831#discussion_r256226539) [](ancestors = 256226539,254097210)